### ISSUE TITLE
Fat arrow binding and the 'this' context

### ISSUE THREAD

**Issue by gavinmcdermott** [harc/ohm#196]
Diving into `ohm` today and had a brief question about the ES6 fat arrow and losing my `this` context within the call to grammar's `addOperation` method. Here's the gist:

In ES5-land...

```js
// Using ES5 syntax, `this` is the object we expect (has the properties, etc.)
const sem = grammar.createSemantics().addOperation('toJS', {
  Number: function (arg) {
    return parseInt(this.sourceString, 10)
  }
})
```

Meanwhile in ES6-land...

```js
// Using ES6 syntax, `this` is undefined; `arg` provides the expected context
const sem = grammar.createSemantics().addOperation('toJS', {
  Number: (arg) => {
    return parseInt(arg.sourceString, 10)
  }
})
```

As I understand it, the ES6 `=>` binds the expected context to inner function. So the ES6 behavior above was unexpected (I could be missing something fundamental though...). Thanks in advance for any pointers here. Also thanks for the great work—been following this for a while and now digging in!

**Comment by pdubroy** [comment]
In ES6 arrow functions, the value of `this` inside the function is the same as its meaning outside the function. Basically, it's equivalent to the following:

```js
Number: (function(arg) {
  ...
}).bind(this)
```

This means that arrow functions aren't really appropriate for defining methods inside an object literal (as is the case here) because methods need a dynamic `this`.

The two examples you've given above are doing different things. In Ohm, the arguments to a semantic action always represent the child nodes of `this`. So in your ES6 example, you are getting the `sourceString` property of the _child_ of the Number node. Depending on your grammar, this might have the same value as the property of the Number node (`this.sourceString` in your first example) -- but it's a different object.

Long story short, when writing semantic actions in ES6, you probably want to use the new [method shorthand](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions) rather than arrow functions:

```js
const sem = grammar.createSemantics().addOperation('toJS', {
  Number(arg) {
    return parseInt(arg.sourceString, 10)
  }
})
```

**Comment by gavinmcdermott** [comment]
@pdubroy: Thank you for the follow up! Much appreciated.
